File: services/content-service/src/main/java/dev/animedia/contentservice/domain/shared/exception/AppException.java
package dev.animedia.contentservice.domain.shared.exception;

import java.util.logging.Level;
import java.util.logging.Logger;

public class AppException extends RuntimeException {
    private final AppExceptionStatus status;
    private final String code;
    private final String[] args;
    private static final Logger LOGGER = Logger.getLogger(AppException.class.getName());

    public AppException(AppExceptionStatus status, String code, String... args) {
        LOGGER.log(Level.WARNING, "App exception: {0}", String.format("Code: %s;", code));
        this.status = status;
        this.code = code;
        this.args = args;
    }

    public AppExceptionStatus getStatus() {
        return status;
    }

    public String getCode() {
        return code;
    }

    public String[] getArgs() {
        return args;
    }
}

File: services/social-service/src/core/Domain/Shared/Exception/AppException.php
<?php

namespace Core\Domain\Shared\Exception;

use RuntimeException;

class AppException extends RuntimeException
{
    /**
     * @param AppExceptionStatus $appExceptionStatus
     * @param string $errorCode
     * @param string[] $args
     */
    public function __construct(
        private readonly AppExceptionStatus $appExceptionStatus = AppExceptionStatus::INTERNAL_ERROR,
        private readonly string $errorCode = 'app.internal_error',
        private readonly array $args = []
    ) {
        parent::__construct(message: "AppException: $errorCode");
    }

    /**
     * @return AppExceptionStatus
     */
    public function getAppExceptionStatus(): AppExceptionStatus
    {
        return $this->appExceptionStatus;
    }

    /**
     * @return string
     */
    public function getErrorCode(): string
    {
        return $this->errorCode;
    }

    /**
     * @return string[]
     */
    public function getArgs(): array
    {
        return $this->args;
    }
}
